crates/core/src/execution_context.rs

```diff
@@ -1,17 +1,125 @@
+                self.rdb_num_rows_fetched += val;
+            }
+        }
+    }
+}
+
+impl BufferMetric {
+    pub fn new(table_id: TableId, table_name: &str) -> Self {
+        Self {
+            table_id,
+            rdb_num_index_seeks: 0,
+            rdb_num_keys_scanned: 0,
+            rdb_num_rows_fetched: 0,
+            cache_table_name: table_name.to_string(),
+        }
+    }
+}
+
+#[derive(Default, Clone)]
+pub struct Metrics(Vec<BufferMetric>);
+impl Metrics {
+    pub fn new() -> Self {
+        Self(Vec::new())
+    }
+
+    pub fn add_metric(&mut self, table_id: TableId, table_name: &str) {
+        self.0.push(BufferMetric::new(table_id, table_name));
+    }
+
+    pub fn inc_by(&mut self, table_id: TableId, ty: MetricType, val: u64) {
+        if let Some(metric) = self.0.iter_mut().find(|x| x.table_id == table_id) {
+            metric.inc_by(ty, val);
+        }
+    }
+
+    pub fn table_exists(&self, table_id: TableId) -> bool {
+        self.0.iter().any(|x| x.table_id == table_id)
+    }
+
+    fn flush(&mut self, workload: &WorkloadType, database: &Address, reducer: &str) {
+        self.0.iter().for_each(|metric| {
+            DB_METRICS
+                .rdb_num_index_seeks
+                .with_label_values(
+                    workload,
+                    database,
+                    reducer,
+                    &metric.table_id.0,
+                    &metric.cache_table_name,
+                )
+                .inc_by(metric.rdb_num_index_seeks);
+
+            DB_METRICS
+                .rdb_num_keys_scanned
+                .with_label_values(
+                    workload,
+                    database,
+                    reducer,
+                    &metric.table_id.0,
+                    &metric.cache_table_name,
+                )
+                .inc_by(metric.rdb_num_keys_scanned);
+
+            DB_METRICS
+                .rdb_num_rows_fetched
+                .with_label_values(
+                    workload,
+                    database,
+                    reducer,
+                    &metric.table_id.0,
+                    &metric.cache_table_name,
+                )
+                .inc_by(metric.rdb_num_keys_scanned);
+        });
+    }
+}
 
 /// Represents the context under which a database runtime method is executed.
 /// In particular it provides details about the currently executing txn to runtime operations.
 /// More generally it acts as a container for information that database operations may require to function correctly.
-#[derive(Default)]
+#[derive(Default, Clone)]
 pub struct ExecutionContext<'a> {
     /// The database on which a transaction is being executed.
     database: Address,
     /// The reducer from which the current transaction originated.
     reducer: Option<&'a str>,
     /// The type of workload that is being executed.
     workload: WorkloadType,
+    /// The Metrics to be reported for this transaction.
+    pub metrics: RefCell<Metrics>,
```

Comment:
I'm guessing RefCell is to avoid changing all of the call sites from `ctx: &ExecutionContext` to `ctx: &mut ExecutionContext`?

Reply:
yes